### PR TITLE
AOB-55: Allow users to setup her own timezone

### PR DESCRIPTION
**Pull request by momoss**
**Description (for Contributor and Core Developer)**

Add timezone field to UserForm

screen shot 2018-03-07 at 17 35 34

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | OK
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | OK
| Review and 2 GTM                  | OK
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: CHANGELOG-2.2.md [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 ### Constructors
 
+- AOB-55: Add method `getTimezone` and `setTimezone` to `Pim\Bundle\UserBundle\Entity\UserInterface`
```
Comment (skeleton) [inline CHANGELOG-2.2.md]: no, this section is on 2.2.0-alpha, create a new section 2.2.x

File: src/Pim/Bundle/UserBundle/Resources/config/doctrine/User.orm.yml [hunk 2]
```diff
@@ -95,6 +95,10 @@ Pim\Bundle\UserBundle\Entity\User:
             type: boolean
             options:
                 default: false
+        timezone:
+            type: string
+            length: 255
```
Comment (skeleton) [inline src/Pim/Bundle/UserBundle/Resources/config/doctrine/User.orm.yml]: why 255?

File: src/Pim/Bundle/UserBundle/Resources/config/doctrine/User.orm.yml [hunk 3]
```diff
@@ -95,6 +95,10 @@ Pim\Bundle\UserBundle\Entity\User:
...
+            nullable: true
```
Comment (skeleton) [inline src/Pim/Bundle/UserBundle/Resources/config/doctrine/User.orm.yml]: no sure we should have a null for timezone, we should have a default (UTC)

**Comment by skeleton** [comment]
don't forget the migration script

File: src/Pim/Bundle/UserBundle/Resources/config/doctrine/User.orm.yml (hunk 2 above)
Comment (momoss) [reply src/Pim/Bundle/UserBundle/Resources/config/doctrine/User.orm.yml]: After check the max timezone length is 32

File: src/Pim/Bundle/UserBundle/Entity/UserInterface.php [hunk 4]
```diff
@@ -528,4 +528,18 @@ public function getPhone(): ?string;
      * @return UserInterface
      */
     public function setPhone(?string $phone): UserInterface;
+
+    /**
+     * Return the User timezone
+     *
+     * @return string|null
```
Comment (skeleton) [inline src/Pim/Bundle/UserBundle/Entity/UserInterface.php]: string

File: src/Pim/Bundle/UserBundle/Entity/User.php [hunk 5]
```diff
@@ -1044,4 +1049,22 @@ public function setPhone(?string $phone): UserInterface
 
         return $this;
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTimezone()
```
Comment (skeleton) [inline src/Pim/Bundle/UserBundle/Entity/User.php]: : string

File: src/Pim/Bundle/UserBundle/Entity/UserInterface.php [hunk 6]
```diff
@@ -528,4 +528,18 @@ public function getPhone(): ?string;
...
+     */
+    public function getTimezone();
```
Comment (skeleton) [inline src/Pim/Bundle/UserBundle/Entity/UserInterface.php]: : string

File: src/Pim/Bundle/UserBundle/Entity/UserInterface.php [hunk 7]
```diff
@@ -528,4 +528,18 @@ public function getPhone(): ?string;
...
+
+    /**
+     * @param ?string $timezone
```
Comment (skeleton) [inline src/Pim/Bundle/UserBundle/Entity/UserInterface.php]: string

File: src/Pim/Bundle/UserBundle/Resources/config/doctrine/User.orm.yml [hunk 8]
```diff
@@ -99,6 +99,9 @@ Pim\Bundle\UserBundle\Entity\User:
             type: string
             nullable: true
             length: 20
+        timezone:
+            type: string
+            length: 30
```
Comment (skeleton) [inline src/Pim/Bundle/UserBundle/Resources/config/doctrine/User.orm.yml]: can you add `nullable: false`, I never know what is the default value of nullable, I prefer to have an explicite configuration

File: features/user/edit_user.feature [hunk 9]
```diff
@@ -55,3 +55,14 @@ Feature: Edit a user
     When I am logged in as "Mary"
     And I edit the "admin" user
     Then I should see the text "Edit user - John Doe"
+
+  Scenario: Successfully setup user timezone
+    When I edit the "Peter" user
```
Comment (phaseinducer) [inline features/user/edit_user.feature]: Given

File: features/user/edit_user.feature [hunk 10]
```diff
@@ -55,3 +55,14 @@ Feature: Edit a user
...
+    And I visit the "Interfaces" tab
+    Then I should see the text "UTC"
+    And I fill in the following information:
```
Comment (phaseinducer) [inline features/user/edit_user.feature]: When

File: features/user/edit_user.feature [hunk 11]
```diff
@@ -55,3 +55,14 @@ Feature: Edit a user
...
+      | Timezone | Paris |
+    Then I save the user
```
Comment (phaseinducer) [inline features/user/edit_user.feature]: And

File: features/user/edit_user.feature [hunk 12]
```diff
@@ -55,3 +55,14 @@ Feature: Edit a user
...
+    And I reload the page
+    Then I visit the "Interfaces" tab
```
Comment (phaseinducer) [inline features/user/edit_user.feature]: And

File: features/user/edit_user.feature [hunk 13]
```diff
@@ -55,3 +55,14 @@ Feature: Edit a user
...
+    And I should see the text "Paris CET (UTC+01:00)"
```
Comment (phaseinducer) [inline features/user/edit_user.feature]: Then

File: src/Pim/Bundle/UserBundle/Form/Type/UserType.php [hunk 14]
```diff
@@ -287,6 +287,24 @@ protected function setDefaultUserFields(FormBuilderInterface $builder)
...
                 ]
+            )
+            ->add(
+                'timezone',
+                TimezoneType::class,
+                [
+                    'label'    => 'user.timezone',
+                    'required' => false,
+                    'choice_label' => function ($timezone, $key) {
+                        $currentDateTime = new \DateTime("now", new \DateTimeZone($timezone));
```
Comment (phaseinducer) [inline src/Pim/Bundle/UserBundle/Form/Type/UserType.php]: simple quotes

File: upgrades/schema/Version_2_2_20180308081108_add_user_timezone.php [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+ * Auto-generated Migration: Please modify to your needs!
+ */
+class Version_2_2_20180308081108_add_user_timezone extends AbstractMigration
+{
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
+    {
+        $this->addSql('ALTER TABLE oro_user ADD timezone VARCHAR(30) NOT NULL');
```
Comment (phaseinducer) [inline upgrades/schema/Version_2_2_20180308081108_add_user_timezone.php]: If the field is mandatory, the migration should also set the default value.

**Pull request merged**